pkg/controllers/mcs/mcs_serviceexport_controller.go

```diff
@@ -0,0 +1,235 @@
+// generates ServiceExport and propagate it to the source clusters.
+// Note: This controller only focus on  multiClusterService of type CrossCluster.
+type MultiClusterServiceController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+func (c *MultiClusterServiceController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling MultiClusterService %s.", req.NamespacedName.String())
+
+	mcs := &networkingv1alpha1.MultiClusterService{}
+	if err := c.Client.Get(ctx, req.NamespacedName, mcs); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if err := c.reconcileServiceExport(ctx, mcs); err != nil {
+		klog.Errorf("Failed to handle ServiceExport for MultiClusterService(%s), error: %v", req.NamespacedName.String(), err)
+		return controllerruntime.Result{}, err
+	}
+	if !mcs.DeletionTimestamp.IsZero() || !mcs.Spec.HasExposureTypeCrossCluster() {
+		return c.removeFinalizer(mcs)
+	}
+	return c.ensureFinalizer(mcs)
+}
+
+func (c *MultiClusterServiceController) deriveServiceExportFromMultiClusterService(
+	ctx context.Context, mcs *networkingv1alpha1.MultiClusterService, ownSvc *corev1.Service) error {
+	oldSvcExport := &mcsv1alpha1.ServiceExport{}
+	err := c.Client.Get(ctx, types.NamespacedName{Name: mcs.Name, Namespace: mcs.Namespace}, oldSvcExport)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		newSvcExport := &mcsv1alpha1.ServiceExport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: mcs.Namespace,
+				Name:      mcs.Name,
+				Annotations: map[string]string{
+					networkingv1alpha1.ResourceKindMultiMultiClusterServiceUsedBy: string(networkingv1alpha1.ExposureTypeCrossCluster),
+				},
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(ownSvc, corev1.SchemeGroupVersion.WithKind("Service")),
+				},
+			},
+		}
+		err = c.Client.Create(ctx, newSvcExport)
+		if err != nil {
+			msg := fmt.Sprintf("Create derived ServiceExport(%s/%s) failed, Error: %v", newSvcExport.Namespace, newSvcExport.Name, err)
+			klog.Errorf(msg)
+			c.EventRecorder.Eventf(mcs, corev1.EventTypeWarning, events.EventSyncDerivedServiceExportFailed, msg)
+			return err
+		}
+		msg := fmt.Sprintf("Sync the serviceExport of MultiClusterService(%s/%s) successful", mcs.Namespace, mcs.Name)
+		klog.V(4).Infof(msg)
+		c.EventRecorder.Eventf(mcs, corev1.EventTypeNormal, events.EventSyncDerivedServiceExportSucceed, msg)
+		return nil
+	}
+	if addCrossClusterToUsedByAnnotationIfNeed(oldSvcExport) {
+		oldSvcExport.OwnerReferences = append(oldSvcExport.OwnerReferences, *metav1.NewControllerRef(ownSvc, corev1.SchemeGroupVersion.WithKind("Service")))
```

Comment:
Is this line of operation unnecessary?

Reply:
Here we mainly deal with the situation where users manually create `serviceExport` before creating MCS. Then, when users manually delete SVC, the serviceExport will be automatically deleted as well.

I am also hesitating whether this line of code is necessary or not.

Reply:
Thanks, I get it. Keeping this line seems fine.